beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java

```diff
@@ -302,11 +317,16 @@ void stop() {
   }
 
   void stop(boolean nack) {
+    nackAllIncoming = nack;
     stopped.set(true);
     if (nack) {
       List<Pair<StreamElement, UnifiedContext>> drop = new ArrayList<>();
       queue.drainTo(drop);
       drop.stream().map(Pair::getSecond).filter(Objects::nonNull).forEach(UnifiedContext::nack);
+      Optional.ofNullable(lastWrittenContext).ifPresent(UnifiedContext::nack);
```

Comment:
isn't nacking handled in that busy waiting loop above (putToQueue)?

Reply:
Actually no. There are quite many states and code-paths, we must make sure that we nack all read-but-not-consumed data. `putToQueue` might not be called at all or it might be called before call to `stop()`, there are many corner cases.

Reply:
related to https://github.com/datadrivencz/proxima-platform/pull/460#discussion_r533207074